tritium-metrics-jvm/src/main/java/com/palantir/tritium/metrics/jvm/JvmMetrics.java

```diff
@@ -77,6 +79,37 @@ private static void registerClassLoading(InternalJvmMetrics metrics) {
         metrics.classloaderUnloaded(classLoadingBean::getUnloadedClassCount);
     }
 
+    private static void registerThreads(InternalJvmMetrics metrics) {
+        ThreadMXBean threads = ManagementFactory.getThreadMXBean();
+        metrics.threadsCount(threads::getThreadCount);
+        metrics.threadsDaemonCount(threads::getDaemonThreadCount);
+        ThreadDeadlockDetector deadlockDetector = new ThreadDeadlockDetector(threads);
+        metrics.threadsDeadlockCount(
+                () -> deadlockDetector.getDeadlockedThreads().size());
+        metrics.threadsNewCount(() -> threadsByState(threads, Thread.State.NEW));
+        metrics.threadsRunnableCount(() -> threadsByState(threads, Thread.State.RUNNABLE));
+        metrics.threadsBlockedCount(() -> threadsByState(threads, Thread.State.BLOCKED));
+        metrics.threadsWaitingCount(() -> threadsByState(threads, Thread.State.WAITING));
+        metrics.threadsTimedWaitingCount(() -> threadsByState(threads, Thread.State.TIMED_WAITING));
+        metrics.threadsTerminatedCount(() -> threadsByState(threads, Thread.State.TERMINATED));
+    }
+
+    private static int threadsByState(ThreadMXBean threads, Thread.State requestedState) {
+        // max-depth zero to avoid creating stack traces, we're only interested in high level metadata
+        final ThreadInfo[] loadedThreadInfo = threads.getThreadInfo(threads.getAllThreadIds(), 0);
```

Comment:
this might be premature optimization, but should we share the thread info across all thread state gauges?

Reply:
I'd thought about this, but I'm not sure how helpful it would be in practice. We can memoize a supplier for the array of threads, but that makes it less likely the array will be quickly garbage collected, though I'm not confident escape analysis will be able to help us out much in this case.

There's no guarantee that all state gauges will be called in quick succession, so sharing the underlying thread info can be a bit complicated to do cleanly.

Reply:
I suppose memoizing an enum map of state to count would work nicely

Reply:
Updated.